### ISSUE TITLE
[1.12.2] Dramatic Armor Trait oddity.

### ISSUE THREAD

**Issue by SonicX8000** [TheIllusiveC4/ConstructsArmory#177]
Minecraft: 1.12.2
Forge: 14.23.5.2838

Mods: 3
conarm-1.12.2-1.2.4
Mantle-1.12-1.3.3.55
TConstruct-1.12.2-2.12.0.157

-----

The Dramatic Armor Trait says that if you were to take a fatal hit, there's 10% chance per Armor Piece that the damage will be reduced to 0 and you'll recover 1 HP from the attack.

It seems that if the damage is high enough...  it won't save you. For testing I used a command to give myself the 'instant_damage 3' potion effect for 1 second. This hits me for 48 damage according to ToroHUD's damage particle.

Game difficulty is on Hard. Wearing 4 Armor Pieces with the 'Dramatic' Trait which should be a 40% chance to not get KO'd from fatal damage. My max health is 20.

*Test 1*
I used the command and Dramatic proc'd, I'm left with 10 Health. ToroHUD showed a 10 damage particle but a 2 healing particle.

I ran the command again after the first and Dramatic proc'd again and I'm left with 3 Health. ToroHUD showed a 10 damage particle but a 2 healing particle again.

At this point, either I couldn't get it to proc'd again or maybe it didn't heal me enough to survive the damage. If Dramatic didn't proc... I take 48 damage.

*Test 2*
For this one... I used 'instant_damage 9' for 1 second which hits me for 3072 damage. Still wearing the same armor as before.

I ran the command and it killed me... however ToroHUD showed a 614 damage particle but with a 2 healing particle so I guess Dramatic proc'd but because the damage was quite high... it didn't save me. If Dramatic didn't proc... I take 3072 damage.

-----

Perhaps maybe 'Dramatic' could be changed so it 'revives' you should you die but with 1 HP. Much like how the Totem of Undying works or like the Focus Sash from Dark Utilities where you're left with 1 HP if you took a fatal hit from full health.

Unsure if this works with Scaling Health where you can have the damage scale depending on it's difficulty. Like... for example, if Dramatic was taking into account for the original damage but not the scaled damage.

*Ex. My max health is 30 and I take 20 damage but it's been scaled up by 2x so I take 40 damage but Dramatic only sees the original 20 damage, meaning it wouldn't proc despite dying from the damage.*

**Comment by kierajreed** [comment]
This is a very good idea. It should block all of the damage and set your health to 1 (half heart).

**Comment by SonicX8000** [comment]
@SkyHawkB 
With what's said in the description for the Traits & Modifiers Guide it should do that but apparently not all of the damage is being blocked, hence why you can still get KO'd from high damaging hits despite the trait saving you.

Changing it to revive you with 1 HP would fit in with the "Saving you in a pinch" tooltip since you died but the trait proc'd so you survive the attack. Could include a message to let the player know that they somehow survived the deadly attack or something.

**Comment by SonicX8000** [comment]
Ah, I should mention that if going with the 1 HP revival thing... might want to ensure that it doesn't mistakenly proc a Totem of Undying or something. Maybe depends with the priority or something? Unless you can force set HP to 1 if the damage was fatal.

**Comment by SonicX8000** [comment]
Gave the update a go. The Dramatic Trait now sets the damage to 0 if it proc'd and heals you for 2 health instead. It doesn't appear to proc revival items if taking fatal damage nor the Focus Sash from Dark Utilities if taking fatal damage from full health. Nicely done.

**Comment by SonicX8000** [comment]
Sorry for another post but the Dramatic Trait doesn't account for the player having Absorption, meaning that if they don't have enough Health survive a fatal attack but they have enough absorption... the damage will still be reduced to 0 and heal the player for 2 health even though the attack wasn't fatal.

My current health is 20, I give myself the Instant Damage III Potion effect which hits me for 24 damage. For testing purposes I give myself 100 Absorption which gives me 120 Overall Health.

***What happens:***
The Dramatic Trait procs as it sees me only having 20 Health despite having 100 absorption. The damage taken is reduced to 0 and heals me for 2 health even though the attack wasn't fatal.

***What I was expecting:***
The Dramatic Trait not proccing since there was enough Overall Health to survive the damage. Leaving me with 76 Absorb & 20 Health. It should only proc if there wasn't enough Health/Absorb to withstand fatal damage.

If I only had 20 Health + 4 Absorb, the trait would have a chance to proc since 24 damage taken is enough to KO me. Same thing if having 10 Health & 14 Absorb but if I had 10 Health & 20 Absorb... the trait will not proc and I'll be left with 6 Health & 0 Absorb.

**Comment by SonicX8000** [comment]
Some more info to add on... it also doesn't account for certain damage reductions, such as Armor, the 'Resistance' Modifier, Shielding Trait, Protection Enchantments. It does appear to account for Modded Damage reductions but I'll need to test that further.

Health is at 20. Used the Instant Damage III Potion which deals 24 damage.

Resistance VIII on 4 pieces of Armor, reduces this damage to 9, which still proc's the Dramatic Trait as it thinks I took 24 damage.

Diamond Helm & Diamond Body with Protection IV and 2 pieces of Resistance VIII Armor with the Dramatic trait. Cuts down the magic damage to 11. Still procs the Dramatic trait as it thinks I took 24 damage.

Resistance Potion effect, for this test I gave myself Resistance IV. This cuts down the damage taken from 24 to 5. Dramatic trait still procs as it thinks I took 24 damage.

'Shielding' Trait which reduces magic damage. Since Instant Damage is magic damage this trait will reduce the damage taken to 14 with 4x Shielding. Dramatic Trait still procs as it see me taking 24 damage.

Angel of Vengeance's Shield of Faith spell which reduces all damage taken by 50%. This reduces the damage taken to 12. Dramatic Trait hasn't proc'd above 12 Health meaning that it looks to be accounting for the modded damage reduction so it sees that I only take 12 damage.

To see if it accounts for Armor... I'm using a Mechanical User to attack me with an all Evil Infused Cleaver, which deals 25 damage to me. With 17 Armor, this reduces the damage taken to 16, leaving me with 4 Health. Dramatic Trait procs as it thinks I took 25 damage even though it was reduced to 16.

-----

In short... only AoV's damage reduction it accounts for... maybe other modded reductions but reductions via Armor, Protection Enchantments, Resist Potion Effect, certain modifiers like Resistance, certain traits like Shielding... it won't account for them.

*EDIT*
Forgot to test the Resistance potion effect. Added that in.
